platform/mds/mds-web/src/main/java/org/motechproject/mds/web/service/impl/InstanceServiceImpl.java

```diff
@@ -160,12 +167,91 @@ public Object saveInstance(EntityRecord entityRecord, Long deleteValueFieldId) {
                 return service.update(instance);
             }
         } catch (Exception e) {
-            if (entityRecord.getId() == null) {
+            if (entity.getId() == null) {
                 throw new ObjectCreateException(entity.getName(), e);
             } else {
-                throw new ObjectUpdateException(entity.getName(), entityRecord.getId(), e);
+                throw new ObjectUpdateException(entity.getName(), entity.getId(), e);
+            }
+        }
+    }
+
+    /***
+     * Translates an entity from the UI with display-only fields into a savable entity.
+     * The translation allows us to reuse UI widgets and wire protocol for editing
+     * entities with @Discriminated.
+     *
+     * One pseudofield represents the discriminator.
+     * Once a discriminator is selected, the UI will expand the field list using fieldsAdded.
+     *
+     * Encoding of pseudofields is performed by:
+     *      org.motechproject.mds.service.impl.EntityServiceImpl.EncodePseudofields()
+     */
+    private void DecodePseudofields(EntityRecord entityRecord, List<FieldDto> entityFields, List<FieldRecord> fieldrecords, EntityDto[] entityTarget, Long deleteValueFieldId)
+    {
+        EntityDto entity1 = getEntity(entityRecord.getEntitySchemaId());
+        EntityDto entity2 = DecodeEntityFromPfds(entityRecord, entity1, deleteValueFieldId);
+        boolean isDiscriminated = entity2 != null;
+        if(isDiscriminated) {
+            validateCredentials(entity2);
+            validateNonEditableProperty(entity2);
+        } else {
+            validateCredentials(entity1);
+            validateNonEditableProperty(entity1);
+
+        }
+        entityTarget[0] = isDiscriminated ? entity2 : entity1;
+        if(isDiscriminated) {
+            entityFields.addAll(getEntityFieldsByClassName(entityTarget[0].getClassName()));
+            fieldrecords.addAll(DecodeFieldlistFromPfds(entityFields, entityRecord, entityTarget[0]));
+            for(FieldDto fd : entityTarget[0].getFieldsAdded()) {
+                entityFields.add(fd);
             }
+        } else {
+            entityFields.addAll(getEntityFields(entityRecord.getEntitySchemaId()));
+            fieldrecords.addAll(entityRecord.getFields());
+        }
+    }
+
+    private EntityDto DecodeEntityFromPfds(EntityRecord entityRecord, EntityDto entity, Long deleteValueFieldId)
+    {
+        List<FieldRecord> entityFields = entityRecord.getFields();
+
+        FieldRecord fdSubclass = null;
+        for(FieldRecord fd : entityFields) {
+            if("subclass".equals(fd.getName()))
```

Comment:
If this is a new field we add to entities, the name should get placed in some sort of constant